misc/python/materialize/feature_benchmark/executor.py

```diff
@@ -73,6 +78,53 @@ def Kgen(self, topic: str, args: List[str]) -> Any:
     def DockerMem(self) -> int:
         return self._composition.mem("materialized")
 
+    def Messages(self) -> Optional[int]:
+        """Return the sum of all messages in the system from mz_internal.mz_message_counts_per_worker"""
+
+        def one_count(e: Docker) -> Optional[int]:
+            result = e._composition.sql_query(
+                dedent(
+                    """
+                    SELECT SUM(sent) as cnt
+                    FROM
+                        mz_internal.mz_message_counts_per_worker mc,
+                        mz_internal.mz_dataflow_channel_operators_per_worker c
+                    WHERE
+                        c.id = mc.channel_id AND
+                        c.worker_id = mc.from_worker_id AND
+                        from_operator_id IN (
+                            SELECT dod.id
+                            FROM mz_internal.mz_dataflow_operator_dataflows dod
+                        )
+                    """
+                )
+            )
+            if len(result) == 0:
+                return None
+            elif result[0][0] is None:
+                return None
+            else:
+                return int(result[0][0])
+
+        # Loop until the message count converges
+        prev_count: Optional[int] = None
+        for i in range(50):
+            new_count = one_count(self)
+            if new_count is not None and prev_count is not None:
+                pct = (max(prev_count, new_count) / min(prev_count, new_count)) - 1
+                # It has converged
+                if pct < 0.05 and i > 1:
```

Comment:
I am wondering: Why don't we wait for full convergence here, rather than only partial convergence? Aren't we guaranteed to take a measurement only after a scenario is done updating? Or are we concerned about the time it takes to get full convergence?

Reply:
I was getting parasitic readings every now and then. For example, on a system with no user-created dataflows, the `SUM(sent)` may either be `NULL` or `5` for no apparent reason. So I figured it is safer (and faster) to not wait for full convergence.

Reply:
I agree with the faster bit, but not sure about the safer bit as the introspection data are not under the same consistency guarantees as user data. While unlikely, it is theoretically possible that you could get some measurement artifacts.

Regarding the `NULL` vs. `5`: My completed unfounded guess is that the introspection query sometimes may take a bit longer to run and then gets picked up by itself (as it is run through a `oneshot-select` dataflow). It is not reliable to look at messages from one-shot `SELECT`s, and you probably also want to exclude `SUBSCRIBE`s in case you use them for monitoring anything.

So the following condition in the query's `WHERE` clause could be refined as in:

```
    from_operator_id IN (
        SELECT dod.id
        FROM mz_internal.mz_dataflow_operator_dataflows dod
        WHERE dod.dataflow_name NOT LIKE '%oneshot-select%'
          AND dod.dataflow_name NOT LIKE '%subscribe%'
    )
```

Reply:
I have modified the SQL query as you suggested. My intention is to merge this as is, but in the future refactor the collection of the message count.

Right now the system focuses on converging the wallclock measurement, meaning that the workload may run many times. The message count is collected for every such measurement, hence the slow-down and the desire not wait for full convergence.

In a future refactor I hope to decouple the two things -- run the workload separately for wallclock measurements and then only once for message count measurements, giving it plenty of time to converge now that it will only be run once.